Made several changes to min-cut partitioner that allow it to recompute     more things

Chillee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122689
* #122581
* __->__ #121692
* #122688
* #122686



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

Perf results
<img width="862" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/8d44e633-8941-46a6-8e7d-806330a8c890">

Chillee committed:
made some minor cleanups/improvements to partitioner

[ghstack-poisoned]

Chillee committed:
Update on "made some minor cleanups/improvements to partitioner"




[ghstack-poisoned]

Chillee committed:
Update on "made some minor cleanups/improvements to partitioner"




[ghstack-poisoned]

Chillee committed:
Update on "made some minor cleanups/improvements to partitioner"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

Chillee committed:
Update on "made some minor cleanups/improvements to partitioner"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "made some minor cleanups/improvements to partitioner"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "made some minor cleanups/improvements to partitioner"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "made some minor cleanups/improvements to partitioner"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "made some minor cleanups/improvements to partitioner"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee commented on torch/_functorch/partitioners.py:
```diff
@@ -379,31 +379,6 @@ def pointwise_ops():
 
     return ops
 
-def get_depth(node, depth_map):
```
python unfortunately has a very low recursion limit (1000), so rewrote into a non-recursive algorithm.

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison commented on torch/_functorch/partitioners.py:
```diff
@@ -733,10 +722,10 @@ def classify_nodes(joint_module):
     if compiler == "inductor":
-        default_recomputable_ops += [prims.div, prims.convert_element_type, aten.clone, aten._to_copy, aten.full_like, prims.var, prims.sum, aten.var, aten.std, prims.broadcast_in_dim, aten.select, aten.permute, aten._unsafe_view, aten.view, aten.expand, aten.slice, aten.reshape, aten.broadcast_tensors, aten.scalar_tensor, aten.ones, aten.new_zeros, aten.lift_fresh_copy, aten.arange, aten.triu, a ...
-        view_ops += [aten.view, aten.slice, aten.permute, aten.t, prims.broadcast_in_dim, aten.expand, aten.as_strided]
+        default_recomputable_ops += [prims.div, prims.convert_element_type, aten.clone, aten._to_copy, aten.full_like, prims.var, prims.sum, aten.var, aten.std, prims.broadcast_in_dim, aten.select, aten._unsafe_view, aten.view, aten.expand, aten.slice, aten.reshape, aten.broadcast_tensors, aten.scalar_tensor, aten.ones, aten.new_zeros, aten.lift_fresh_copy, aten.arange, aten.triu, aten.var_mean,  ...
```
Can we use tags here ? It's not surprising that are missing operators.

Chillee replied:
I don't think we have a view tag right?

peterbell10 replied:
If you had a function `is_recomputable(node)` then it could check 
```python
get_node_storage(node) == get_node_storage(node.args[0])
```

eellison commented on torch/_functorch/partitioners.py:
```diff
@@ -733,10 +722,10 @@ def classify_nodes(joint_module):
-        default_recomputable_ops += [prims.div, prims.convert_element_type, aten.clone, aten._to_copy, aten.full_like, prims.var, prims.sum, aten.var, aten.std, prims.broadcast_in_dim, aten.select, aten.permute, aten._unsafe_view, aten.view, aten.expand, aten.slice, aten.reshape, aten.broadcast_tensors, aten.scalar_tensor, aten.ones, aten.new_zeros, aten.lift_fresh_copy, aten.arange, aten.triu, a ...
-        view_ops += [aten.view, aten.slice, aten.permute, aten.t, prims.broadcast_in_dim, aten.expand, aten.as_strided]
+        default_recomputable_ops += [prims.div, prims.convert_element_type, aten.clone, aten._to_copy, aten.full_like, prims.var, prims.sum, aten.var, aten.std, prims.broadcast_in_dim, aten.select, aten._unsafe_view, aten.view, aten.expand, aten.slice, aten.reshape, aten.broadcast_tensors, aten.scalar_tensor, aten.ones, aten.new_zeros, aten.lift_fresh_copy, aten.arange, aten.triu, aten.var_mean,  ...
+        view_ops += [aten.view, aten.slice, aten.t, prims.broadcast_in_dim, aten.expand, aten.as_strided, aten.permute]
```
Is there a better way of maintaining this list too ?

Chillee replied:
I'll add a view tag in a follow-up PR

eellison replied:
that or @peterbell10 suggestion

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

shunting314 commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -339,6 +339,7 @@ def _precompile_config(self, cfg: Config, warm_cache_only_with_cc: Optional[int]
                 "num_warps": compile_meta["num_warps"],
                 "num_stages": compile_meta["num_stages"],
                 "debug": compile_meta["debug"],
+                "enable_fp_fusion": False,
```
This is to disable FMA?

I'm wondering if this should be in a separate PR even though it's just a single line. I know that you recently posted about an numerical issue caused by FMA. But on the other hand, theoretically speaking disabling FMA may have perf hit due to more computation instructions. Do we observe any perf impact?

Chillee replied:
Yeah I'll put it in a separate PR - @peterbell10 is working on that.

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison commented on torch/_functorch/config.py:
```diff
@@ -32,11 +32,32 @@
 
 # Restricts the amount of computation AOTAutograd can do.
-max_dist_from_bw = 3
+max_dist_from_bw = 1000
```
where is this number coming from ? if it's this large, isnt even worth having ?

Chillee replied:
Yeah, it's basically removed now - keeping it for now as a knob in case somebody finds it necessary in the short term.

eellison commented on torch/_functorch/partitioners.py:
```diff
@@ -452,7 +426,11 @@ def reordering_to_mimic_autograd_engine(gm):
+        if node.op == 'placeholder':
+            depths[node] = 0
+        else:
+            depths[node] = max([depths[arg] for arg in node.all_input_nodes if isinstance(arg, torch.fx.Node)], default=0)
```
`for arg in node.all_input_nodes if isinstance(arg, torch.fx.Node)`

dont think you need the isinstance here with `node.all_input_nodes`

eellison commented on torch/_functorch/partitioners.py:
```diff
@@ -474,7 +452,8 @@ def insert_node_in_graph(node):
                 minimum_order = order[user]
                 first_node_in_bwd = user
-    assert first_node_in_bwd is not None
+    if first_node_in_bwd is None:
+        return gm
```
why is this ?

Chillee replied:
Fixes a bug in existing impl - if gradInp does not depend on gradOut then this will not detect any nodes in the bacwkards.

Added a comment.

eellison commented on torch/_functorch/partitioners.py:
```diff
@@ -718,6 +697,19 @@ def classify_nodes(joint_module):
+    def is_fusible(a, b):
+        # We can perform "memory fusion" into a cat, but cat cannot be a
+        # producer to a fusion
+        if get_aten_target(b) == aten.cat:
+            return True
```
should this be ?
```
if get_aten_target(b) == aten.cat and get_aten_target(a): 
	return False
```

Chillee replied:
The idea is that you can always "fuse" ops into cat, so recomputing an op that feeds into cat will always be free. But this was here before - this is just code movement.

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison commented on torch/_functorch/partitioners.py:
```diff
@@ -749,11 +741,10 @@ def classify_nodes(joint_module):
 
     random_ops = [aten.native_dropout, aten.rand_like, aten.randn_like]
-    compute_intensive_ops = [aten.mm, aten.convolution, aten.convolution_backward, aten.bmm, aten.addmm, aten.upsample_bilinear2d, aten._softmax, aten._softmax_backward_data, aten.native_layer_norm, aten.native_layer_norm_backward, aten.native_batch_norm, aten.native_batch_norm_backward, aten._native_batch_norm_legit, aten._batch_norm_with_update, aten.batch_norm_backward]  # noqa: E501,B950
+    compute_intensive_ops = [aten.mm, aten.convolution, aten.convolution_backward, aten.bmm, aten.addmm, aten._scaled_dot_product_flash_attention, aten._scaled_dot_product_efficient_attention, aten.upsample_bilinear2d]  # noqa: E501,B950
```
batchnorm/softmax/layernorm were able to be removed bc of reduction changes ? how did you decide upsample_bilinear2. 

i think you might want max_pool2d_backward as well

Chillee replied:
They were removed since we never see them nowadays at this point.

This list is not used unless aggressive recompute is turned on. In that case, I think it might make more sense to not recompute things like upsampling. This is more preparation for later changes for more user control for recomputation.

eellison commented on torch/_functorch/partitioners.py:
```diff
@@ -766,64 +757,76 @@ def classify_nodes(joint_module):
+            log.info("materialized backwards: %s %s", node, tuple(node.users))
+            return True
+
+        # Arbitrary hack that sometimes seems to help things. The above
+        # modification appears to have made this heuristic a lot less critical
+        # for performance.
+        if not graph_has_recomputable_ops:
```
which heuristic specifically ? `is_materialized_backwards` ?

Chillee replied:
Yeah. Added a comment noting its obsolescence.

eellison commented on torch/_functorch/partitioners.py:
```diff
@@ -837,15 +840,27 @@ def get_node_weight(node) -> int:
+        # ancestor of this node to the backwards path through this node that
+        # doesn't go through any saved value. If this node is saved, then that
+        # condition is not possible.
+        nx_graph.add_edge("source", node.name + "_in", capacity=math.inf)
```
worth factoring out the inf banning of recomputation into helper ?

Chillee replied:
This is the helper haha.

eellison commented on torch/_functorch/partitioners.py:
```diff
@@ -888,6 +903,86 @@ def get_node_weight(node) -> int:
             nx_graph.add_edge(node.name + "_out", user.name + "_in", capacity=math.inf)
 
+    # todo(chilli): This is the most questionable of the 3 heuristics for banning recompute.
+    # I would look more into it but I'm out of time for now
```
you can prob leave this sentence out haha. that's what are all todos are for lol

eellison commented on torch/_functorch/partitioners.py:
```diff
@@ -888,6 +903,86 @@ def get_node_weight(node) -> int:
+    # backwards pass instead of only relying on whether it's unfusible in the
+    # forwards.
+
+    def find_first_unfusible(start_nodes, max_range):
```
nit: add type annotation, docstring ?

eellison commented on torch/_functorch/partitioners.py:
```diff
@@ -888,6 +903,86 @@ def get_node_weight(node) -> int:
+
+    def find_first_unfusible(start_nodes, max_range):
+        sorted_nodes = []
+        for n in start_nodes:
```
still need to review rest of this file, will do tomorrow

eellison reviewed (commented):
part way through review

Chillee commented on benchmarks/dynamo/timm_models.py:
```diff
@@ -262,12 +263,21 @@ def load_model(
+            'https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/beignets-task-guide.png'
+        ))
+        transforms = create_transform(**data_config, is_training=False)
+        example_inputs = transforms(img).to(device=device).half().unsqueeze(0).repeat(batch_size, *([1]*len(input_size)))
```
Ah I've since removed these changes.

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee commented on torch/_dynamo/utils.py:
```diff
@@ -364,12 +364,12 @@ def setup_compile_debug():
-            inductor=logging.DEBUG,
-            output_code=True,  # this is off by default
-        )
+        # torch._logging.set_logs(
```
Was gonna disable this, but will push to a later PR.

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee commented on torch/_functorch/partitioners.py:
```diff
@@ -921,10 +1022,14 @@ def get_node_weight(node) -> int:
                 joint_module, fw_module, bw_module, len(saved_sym_nodes)
             )
-        bw_module = reordering_to_mimic_autograd_engine(bw_module)
+    bw_module = reordering_to_mimic_autograd_engine(bw_module)
```
This is deliberate btw.

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

shunting314 commented on torch/_functorch/partitioners.py:
```diff
@@ -888,6 +906,89 @@ def get_node_weight(node) -> int:
+            if start_node not in required_fw_nodes:
+                continue
+            fusible = [(start_node.fw_order, start_node)]
+            start_pos = start_node.fw_order
```
not used

shunting314 commented on torch/_functorch/partitioners.py:
```diff
@@ -888,6 +906,89 @@ def get_node_weight(node) -> int:
+                if cur in visited:
+                    continue
+                visited.add(cur)
+                if cur.fw_order > start_order + 100 and len(fusible) == 0:
```
100 here does not necessarily means the chains contains 100 nodes right? Most likely the length of the chain is much less than 100.

shunting314 replied:
I think it's possible to accurately maintain the chain length similar to how `get_depth` works. Not sure if that's necessary though

Chillee replied:
> 100 here does not necessarily means the chains contains 100 nodes right? 

Yeah, I don't think this heuristic is that important actually.

shunting314 commented on torch/_functorch/partitioners.py:
```diff
@@ -888,6 +906,89 @@ def get_node_weight(node) -> int:
+
+    # It's possible this heuristic is obsoleted by the below one - checking that now
+
+    if BAN_IF_LONG_FUSIBLE_CHAINS:
```
A general question about this heuristic, let's say I have a long chain of nodes that I can fuse: n1, n2, ...., n100. Sounds like we should ban the recomputation for n100.

The chain is long, but if we save say n98, then when we recompute n100 in backward pass, we only need recompute n99 and n100. The recomputation cost is not high even though the chain of fusable nodes is long (100 nodes in this case). Then ban recomputation for n100 seems to be not necessary for this specific case? (The example pick an unrealisticaly long chain for discussion purpose)

Chillee replied:
If I understand what you're saying, it's that by banning recomputation for n100, we might prevent other solutions that are good that involve saving n98?

IMO, 2 things:
1. This is kind of a degenerate case, and so I think the precise performance here doesn't matter too much as long as we're not falling into the degenerate case.
2. We do have a heuristic there where we only save it once it narrows down to a "bottleneck".

shunting314 replied:
> If I understand what you're saying, it's that by banning recomputation for n100, we might prevent other solutions that are good that involve saving n98?

Yes. Due to saving n98, recomputing n100 is not that expensive even though the fusible chain is very long.

Chillee commented on torch/_functorch/partitioners.py:
```diff
@@ -379,31 +382,6 @@ def pointwise_ops():
 
     return ops
 
-def get_depth(node, depth_map):
```
Yeah. In this case it was specifically applied to the backwards pass.

Chillee replied:
Yeah. This is also only applied to the backwards graph.

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee commented on torch/_functorch/partitioners.py:
```diff
@@ -888,6 +906,89 @@ def get_node_weight(node) -> int:
+        for start_node in joint_graph.nodes:
+            if start_node not in required_fw_nodes:
+                continue
+            fusible = [(start_node.fw_order, start_node)]
```
I don't think Python has another heap data structure 🤔 https://stackoverflow.com/questions/12749622/understanding-how-to-create-a-heap-in-python

Chillee commented on torch/_functorch/partitioners.py:
```diff
@@ -888,6 +906,89 @@ def get_node_weight(node) -> int:
+    # where we have a long chain of pointwise ops from the beginning to the end
+    # of the model (like say, residual connections)
+
+    # It's possible this heuristic is obsoleted by the below one - checking that now
```
<img width="1085" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/2558b957-47a1-42c1-a865-ce7c838aa7ae">

Chillee commented on torch/_functorch/partitioners.py:
```diff
@@ -888,6 +906,89 @@ def get_node_weight(node) -> int:
+                    continue
+                visited.add(cur)
+                # 100 is arbitrary choice to try and prevent degenerate cases
+                if cur.fw_order > start_order + 100 and len(fusible) == 0:
```
I think right now all the choices are just booleans, so I hesitate on making it an integer.

Chillee commented on torch/_functorch/partitioners.py:
```diff
@@ -888,6 +906,89 @@ def get_node_weight(node) -> int:
+
+    # It's possible this heuristic is obsoleted by the below one - checking that now
+
+    if BAN_IF_LONG_FUSIBLE_CHAINS:
```
We could explore it in follow-up commits but I don't want to retest/isolate the affect of this change.

Chillee commented on torch/_functorch/partitioners.py:
```diff
@@ -888,6 +906,89 @@ def get_node_weight(node) -> int:
+            fw_users = [user for user in used_node.users if user in required_fw_nodes]
+            if len(orders) > 0:
+                first_unfusible_use = find_first_unfusible(fw_users, max(orders))
+                for user in tuple(used_node.users):
+                    if user in required_fw_nodes and user.fw_order > first_unfusible_use and is_fusible(used_node, user):
+                        if user in banned_nodes:
+                            continue
+                        log.info(
+                            "used above/below fusible %s:(%s) -> %s -> %s:(%s)",
+                            used_node, used_node.fw_order, first_unfusible_use, user, user.fw_order
+                        )
+                        ban_recomputation_if_allowed(user)
```
The idea is that if we have something like
```
a: Tensor
b = cos(a)
c = mm(a)
...
d = add(a, c)
```

`b` and `d` cannot get fused together, and so if we choose to recompute `d` we will be reading from `a` multiple times.

It's not a perfect heuristic, but it's what I have now :)

Chillee commented on torch/_functorch/partitioners.py:
```diff
@@ -888,6 +906,89 @@ def get_node_weight(node) -> int:
+                    break
+
+                for user in cur.users:
+                    if user in required_fw_nodes and is_fusible(cur, user) and user not in banned_nodes:
```
It might force a materialization yeah, but since this is only nodes in the forwards pass that's not something that "affects" recomputation.

Chillee commented:
@eellison For the most part, none of the heuristics are explicitly trying to work around Inductor limitations AFAIK. Some of the heuristics definitely aren't perfect, but that's why they're heuristics :)

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

Perf results
<img width="862" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/8d44e633-8941-46a6-8e7d-806330a8c890">


[ghstack-poisoned]

Chillee committed:
Update on "Made several changes to min-cut partitioner that allow it to recompute     more things"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

Perf results
<img width="862" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/8d44e633-8941-46a6-8e7d-806330a8c890">


[ghstack-poisoned]

Chillee commented:
@pytorchbot merge -i

Chillee commented:
@pytorchbot merge -i

Chillee commented:
@pytorchbot merge -i

anijain2305 commented on torch/_functorch/partitioners.py:
```diff
@@ -837,15 +843,33 @@ def get_node_weight(node) -> int:
+    def ban_recomputation_if_allowed(node):
+        # This bans recomputation of the node unless we've been forced not to by
+        # user annotation
+        # NB: "recompute" == 0 means it's not banned from recomputing
```
Not sure about this. We marked the output nodes of AC blocks with "recompute" == 0. Shouldn't we ban recomputation for those nodes?

Chillee replied:
This was correct (in some sense), forward fix here: https://github.com/pytorch/pytorch/pull/122841
